### ISSUE TITLE
Poor PTB test performance?

### ISSUE THREAD

**Issue by D-X-Y** [quark0/darts#50]
Hi Hanxiao,

I train this model https://github.com/quark0/darts/blob/master/rnn/genotypes.py#L33 on PTB. I obtain the similar validation performance (val ppl = 59.0), but the test ppl (61.3) is much higher than the reported results (55.7). Is there any suggestion?

Thanks

**Comment by quark0** [comment]
Something must be wrong. For PTB, it's unlikely to get a higher test ppl than the valid ppl unless using substantially different regularization.

Can you (1) evaluate our pretrained model and see if you can reproduce 55.7; (2) make sure the genotypes for training and for testing are the same; (3) perhaps share with me your log and model?

Also, try running the final evaluation script on the validation set (by replacing the test set as validation set in test.py) and see if you can get 59.0.

**Comment by D-X-Y** [comment]
Hi, Hanxiao @quark0 

I found the problem. I test the model every epoch, but did not use these lines https://github.com/quark0/darts/blob/master/rnn/train.py #L268-#L272 . Though I do not understand these lines, but after I add these lines, the trained model performs well on the test set.
Would you mind to explain these codes? rnn/train.py L268-L272 and L286-L287.

Thanks a lot!

**Comment by quark0** [comment]
For ASGD, we evaluate the averaged model (over the history) rather than a particular snapshot. L268-L272 rewrites the current weights using the averaged weights recorded in the optimizer's state, L274-L279 evaluates the averaged model, and L286-L287 restores the original weights.

For more details, please refer to https://arxiv.org/abs/1708.02182

**Comment by D-X-Y** [comment]
I understand now. Thanks a lot!

**Comment by D-X-Y** [comment]
Hi @quark0 , I re-train DARTS, and obtains 59.45 validation ppl and 57.73 test ppl. Is it ok, or still has something wrong?

**Comment by quark0** [comment]
Why it's different from your previous 59.0? You should be able to get better results by training DARTS_V2 for ~3000 epochs using the latest code and commands given in the README.

**Comment by D-X-Y** [comment]
I train the model by 1600 epochs, 59.0 and 59.45 on the validation set come from two different runs.
These two numbers are obtained by DARTS_V1 (DARTS first order?). I will try DARTS_V2. Thanks a lot!

**Comment by quark0** [comment]
I see. DARTS_V1 was obtained using the old codebase before a bug fix (#21), and will be removed in the future. The current default hyperparameters are tuned wrt DARTS_V2, which should result in ~58.1 valid ppl and ~55.7 test ppl after 3000 epochs of training.

**Comment by D-X-Y** [comment]
@quark0 Thanks a lot!
May I ask how the performance of the model discovered by (DARTS first order) on WT2?

**Comment by quark0** [comment]
We didn't transfer the 1st order cells. I could launch some experiments when I get a chance in the next couple of days and post my results there.

**Comment by D-X-Y** [comment]
@quark0 I appreciate your help!

**Comment by linnanwang** [comment]
Hello there, is it 3000 epochs or 300 epochs? thank you.
